File: src/services/userService.js
import { CowModel } from "../models/cowModel.js";
import { DeviceModel } from "../models/deviceModel.js";
import { SaveZone } from "../models/saveZoneModel.js";
import { UserModel } from "../models/userModel.js";
const createUser = async (req) => {
  const newUser = new UserModel(req.body);
  const savedUser = await newUser.save();

  return savedUser;
};

const getUserByUsername = async (username) => {
  const user = await UserModel.findOne({
    username: username,
  });
  return user;
};
const getAllUser = async () => {
  const users = await UserModel.find();
  return users;
};

const updateByUsername = async (username, updateData) => {
  try {
    const updatedUser = await UserModel.findOneAndUpdate(
      { username: username },
      updateData,
      { new: true, runValidators: true }
    );

    if (!updatedUser) {
      throw new Error(`User with username "${username}" not found.`);
    }

    return updatedUser;
  } catch (err) {
    console.error("Error updating user:", err);
    throw err;
  }
};
const deleteUserById = async (userId) => {
  const user = await UserModel.findById(userId);
  if (user) {
    const username = user.username;

    // Delete the user
    await user.deleteOne();

    // Delete related documents by username
    await Promise.all([
      DeviceModel.deleteMany({ username }),
      CowModel.deleteMany({ username }),
      SaveZone.deleteMany({ username })
    ]);
    console.log(`Deleted user ${username} and related data.`);
  }else{
    console.log("User not found");
  }
}

export const incrementGlobalAddress = async (username) => {
  try {
    const updatedUser = await UserModel.findOneAndUpdate(
      { username: username },
      { $inc: { global_address: 1 } },
      { new: true }
    );
    if (!updatedUser) {
      throw new Error(`User with username "${username}" not found.`);
    }
    return updatedUser;
  } catch (err) {
    console.error("Error updating global address:", err);
    throw err;
  }
};

export default {
  createUser: createUser,
  getUserByUsername: getUserByUsername,
  updateByUsername,
  incrementGlobalAddress,
  getAllUser,
  deleteUserById
};

File: lib/services/user_service.dart
import 'dart:convert';
import 'package:do_an_app/global.dart';
import 'package:do_an_app/models/user_model.dart';
import 'package:http/http.dart' as http;


Future<List<UserModel>?> getAllUser() async {
  try{
    var url = Uri.http(serverUrl, '/user/all');

    var res = await http.get(
      url
    );

    if(res.statusCode == 200){
      List<dynamic> DeviceModelJsons = json.decode(res.body);

      List<UserModel> DeviceModels = [];
      for(final DeviceModelJson in DeviceModelJsons){
        DeviceModels.add(UserModel.fromJson(DeviceModelJson));
      }
      return DeviceModels;
    }else {
      print("getAllDevice failed, status code: ${res.statusCode}");
      return null;
    }
  }catch(err){
    print("getAllDevice failed, error: $err");
    return null;
  }
}

Future<UserModel?> LoginUser(
  String? username,
  String? password,
) async{
  try{
    var url = Uri.http(serverUrl, '/user/login');
    var body = {
      if (username != null) 'username' : username,
      if (password != null) 'password': password,
    };
    var res = await http.post(
      url,
      body: jsonEncode(body),
      headers: {
        'Content-Type': 'application/json',
      },
    );
    if(res.statusCode == 200){
      print("Successfully login user");
      var bodyJson = jsonDecode(res.body);
      return UserModel.fromJson(bodyJson);
    }else{
      print("Login user failed");
      return null;
    }
  }catch(err){
    print("LoginUser failed, error: $err");
    return null;
  }
}
Future<UserModel?> postUser(
  String? username,
  String? password,
  String? fullname,
  int? global_address
) async {
  try{
    var url = Uri.http(serverUrl, '/user/register');
    var body = {
      if (username != null) 'username' : username,
      if (password != null) 'password': password,
      if (fullname != null) 'fullname': fullname,
      if (global_address != null) 'global_address': global_address,
    };
    var res = await http.post(
      url,
      body: jsonEncode(body),
      headers: {
        'Content-Type': 'application/json',
      },
    );
    if(res.statusCode == 201){
      print("Successfully create user");
      var bodyJson = jsonDecode(res.body);
      return UserModel.fromJson(bodyJson);
    }else if(res.statusCode == 424){
      print("Gateway not response");
      var bodyJson = jsonDecode(res.body);
      return UserModel.fromJson(bodyJson);
    }
    else{
      print("postUser failed, status code: ${res.statusCode}");
      return null;
    }
  }catch(err){
    print("postUser failed, error: $err");
    return null;
  }
}
Future<UserModel?> updateUserByUsername(
  String? userId,
  String? username,
  String? fullname,
  int? global_address
) async {
  try {
    var url = Uri.http(serverUrl, '/user');
    var body = {
      if (username != null) 'username': username,
      if (fullname != null) 'fullname': fullname,
      if (global_address != null) 'global_address':global_address
    };
    var res = await http.put(
      url,
      body: jsonEncode(body),
      headers: {
        'Content-Type': 'application/json',
      },
    );
    if (res.statusCode == 200) {
      var bodyJson = jsonDecode(res.body);
      return UserModel.fromJson(bodyJson);
    } else {
      print("updateUserByUsername failed, status code: ${res.statusCode}");
      return null;
    }
  } catch (err) {
    print("updateUserByUsername failed, error: $err");
    return null;
  }
}


Future<UserModel?> getUserByUsername(String username) async {
  try{
    var url = Uri.http(serverUrl, '/user/$username');
    var res = await http.get(
      url
    );
    if(res.statusCode == 200){
      var bodyJson = jsonDecode(res.body);
      return UserModel.fromJson(bodyJson);
    }else {
      print("getUserByUsername failed, status code: ${res.statusCode}");
      return null;
    }
  }catch(err){
    print("getUserByUsername failed, error: $err");
    return null;
  }
}   


Future<int?> getAndIncrementGlobalAddress(
  String? username,
  String? cowId
) async {
  try {
    var url = Uri.http(serverUrl, '/user/global-address/$username');
    var body = {
      if (cowId != null) 'cowId': cowId,
    };
    var res = await http.put(
      url,
      body: jsonEncode(body),
      headers: {
        'Content-Type': 'application/json',
      },
    );
    if (res.statusCode == 200) {
      var responseBody  = jsonDecode(res.body);
      return responseBody['global_address']; 
    } else {
      print("getAndIncrementGlobalAddress failed, status code: ${res.statusCode}");
      return null;
    }
  } catch (err) {
    print("getAndIncrementGlobalAddress failed, error: $err");
    return null;
  }
}


Future<int?> deleteUserById(String userId) async {
  try{
    var url = Uri.http(serverUrl, '/user/delete');
    final body = jsonEncode({'userId': userId});

    var res = await http.delete(
      url,
      headers: {'Content-Type': 'application/json'},
      body: body,
    );
    
    if(res.statusCode == 200){
      print("Delete success");
      return 200;
    }else{
      print("deleteUserById failed, status code: ${res.statusCode}");  
      return res.statusCode;
    }

  }catch(err){
    print("deleteUserById failed, error: $err");
    return null;
  }
}

